flink-tests/src/test/java/org/apache/flink/test/checkpointing/CheckpointStoreITCase.java

```diff
@@ -126,68 +147,39 @@ public HighAvailabilityServices createHAServices(
 
                 @Override
                 public CheckpointRecoveryFactory getCheckpointRecoveryFactory() {
```

Comment:
Would it work to instantiate a `TestingHighAvailabilityServices` with a special `CheckpointRecoveryFactory` instead of creating a sub class of `EmbeddedHaServices`?

Reply:
Not without implementing proper cleanup of leader election services for `TestingHighAvailabilityServices` (otherwise I'll end up with something like `Caused by: java.lang.IllegalStateException: leader retrieval service is already started`).

I think implementing this would basically lead to re-implemention of `EmbeddedHaServices` for purpose of this test.

WDYT?

Reply:
There are actually more sub-classes of EmbeddedHaServices serving the same / similar purpose. All are intended to be used with MiniCluster, probably for the same reason I've mentioned above

Reply:
I was able to get rid of all them, by adding second constructor to `EmbeddedHaServicesWithLeadershipControl`, that allows overriding of `CheckpointRecoveryFactory`. 🎉